docs/docsite/rst/dev_guide/developing_plugins.rst

```diff
@@ -487,33 +487,88 @@ Vars plugins
+To get started developing a vars plugin, create the new file in one of the search paths and configure Ansible to use it as described in :ref:`enable_vars`.
 
-Older plugins used a ``run`` method as their main body/work:
+Since vars plugins can run many times for simple playbooks (see :ref:`using_vars`), they should mitigate their overhead by allowing user configuration of when they run by using the ``vars_plugin_staging`` documentation fragment and caching expensive operations when possible using a global variable.
+
+The main method of a vars plugin is ``get_vars``, which just needs to return a dictionary structure with the variables.
+
+``get_vars`` is called with the parameters:
+
+ * loader: Ansible's DataLoader. The DataLoader can read files, auto-load JSON/YAML and decrypt vaulted data, and cache read files.
+ * path: this is 'directory data' for every inventory source and the current play's playbook directory, so they can search for data in reference to them. ``get_vars`` will be called at least once per available path.
+ * entities: these are host or group names that are pertinent to the variables needed. The plugin will get called once for hosts and again for groups.
+
+.. code-block:: python
+
+   DOCUMENTATION = '''
+       name: ns.col.custom_hostvars
+       version_added: "1.0.0"
+       short_description: Load custom host vars
+       description: Load custom host vars
+       options:
+         stage:
+           ini:
+             - key: stage
+               section: ns_coll_vars_custom_hostvars
+           env:
+             - name: ANSIBLE_VARS_PLUGIN_STAGE
+       extends_documentation_fragment:
+         - vars_plugin_staging
+    '''
+
+   from ansible.errors import AnsibleParserError
+   from ansible.plugins.vars import BaseVarsPlugin
+
+   # global cache
+   FOUND = {}
+
+   def find_variable_source_for_entity(path, entity):
+       pass  # your code here
+
+   class VarsModule(BaseVarsPlugin):
+       def load_variables_from_source(self, loader, source):
+           pass  # your code here
+
+       def get_vars(self, loader, path, entities):
+           variables = {}
+           for entity in entities:
+               try:
+                   cache_key = entity.name + entity.base_type
+               except AttributeError:
+                   raise AnsibleParserError("Supplied entity must be Host or Group, got %s instead" % (type(entity)))
+
+               if cache_key not in FOUND:
+                   FOUND[cache_key] = find_variable_source_for_entity(path, entity)
+
+               variables.update(self.load_variables_from_source(loader, FOUND[cache_key]))
+           return variables
+
+In addition to running many times, a vars plugin is instantiated by default every time it might run. Stateless plugins like ``host_group_vars`` can opt into reusing the same instance throughout a play by setting the ``reuse_instance`` class attribute to ``True``.
```

Comment:
Other attribute name suggestions: ``is_stateless``, ``use_cached_instance``, ``cache_instance``...

Reply:
I'm +0.25 for `is_stateless` over `reuse_instance`, but either is fine with me...

Reply:
I renamed it back to `is_stateless`. It feels like more of a promise on the part of the vars plugin, than a description of the behavior.